sympy/stats/crv.py

```diff
@@ -417,7 +417,7 @@ def moment_generating_function(self, t, **kwargs):
 
     def expectation(self, expr, var, evaluate=True, **kwargs):
         """ Expectation of expression over distribution """
-        if evaluate:
+        if evaluate and expr.is_zero is False:
```

Comment:
This will also fail in some cases. What if `is_zero` gives `None`?

You should check `p` below instead (as I suggested before).

Reply:
> You should check p below instead

Wouldn't `p` suffer with the same problem ? Or is it that `is_zero` doesn't give `None` for `poly` expressions.  Besides isn't the `Integral` doing same thing (mathematically) as the `try - except` block ?

Reply:
Poly are very different from Expr:
```
In [14]: e = x*(x + 1) - x*x - x

In [15]: e
Out[15]: 
   2                
- x  + x⋅(x + 1) - x

In [16]: Poly(e, x)
Out[16]: Poly(0, x, domain='ZZ')
```
The `else` block returns an unevaluated `Integral` which is not what is wanted when `evaluate=True` (compare `integrate` vs `Integral`).

Reply:
> compare integrate vs Integral 

Ah, I see. 

So you meant adding something like : 

```
if p.is_zero:
__return integrate(expr * self.pdf(var), (var, self.set), **kwargs)
```
After p has been evaluated. Will that do?

Reply:
Each time we discuss a particular case like this you should consider whether or not there is a test that covers that case. If there isn't then one should be added.

The change I am commenting on here would have caused an unevaluated integral to be returned whenever `expr.is_zero` gives `None`. Is there a test for that case?

Reply:
Not that I know of. Expectation calculation for `SingleContinuousDistribution` takes in expressions of type distributions which `is_zero` check almost always gives `False` Since the distributions almost always have a value (or rather a variable that contains the distribution which is always regarded as non-zero ). So regardless of how small we make their coefficients or how tangled we make make that equation it gives `is_zero` check as `False`. The other possibility was the expression being exactly zero (as is in the issue). 

If you encounter any such case for which an expression having `is_zero` as `None` is being passed here please mention one here so it could be handled appropriately.

Reply:
I gave an example:
```python
In [1]: e = x*(x + 1) - x*x - x

In [2]: e
Out[2]: 
   2                
- x  + x⋅(x + 1) - x

In [3]: print(e.is_zero)
None
```

Reply:
But currently we have added the `is_zero` check after the conversion into `Poly` for which this particular case also evaluates to `True`. Moreover I think the expression is also being simplified before entering this function. So direct implementation of any such example isn't working.

Reply:
Part of the reason for adding tests is to test the current behaviour of the new code with the changes made in any particular PR. The more important reason though is to test the future behaviour when someone else *changes* the code later. If a future contributor tried to make the change that you did and test `expr.is_zero` then the test failure would show them why that is wrong. If the code that you showed before was able to pass the test suite then that means that there are not enough tests to pick up on this potential problem.

That is why each time you discover a problem with the code in a PR you should not only fix that problem in the code but also *add a test* (unless there already is one). The test makes sure that the code works correctly *now*, and it proves to anyone reviewing the PR that it correctly handles the tested examples. More importantly it ensures that those examples will be handled correctly *in future* even if the code is changed.

Reply:
Thanks, I'll keep this in mind.